### PR TITLE
Refactor base scm functions

### PR DESCRIPTION
**Pull request by tkyi**
In preparation for creating a Bitbucket scm plugin, refactoring some functions in the scm base class:
- Add decorators for url, commit, and author
- Add parsers for url and hook
- Deprecate `scmUrl` in favor of `scmUri` (`github.com:12345:master`)

Blocked by https://github.com/screwdriver-cd/data-schema/pull/67
Related to https://github.com/screwdriver-cd/screwdriver/issues/255

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,23 +20,136 @@ Required parameters:
...
 #### Expected Outcome
 A formatted scm url to be used as a unique key. This function is synchronous.
 
+### parseUrl
+Required parameters:
+
+| Parameter        | Type  |  Description |
+| :-------------   | :---- | :-------------|
+| config             | Object | Configuration Object |
+| config.checkoutUrl | String | Checkout url for a repo to parse |
```
Comment (FenrirUnbound) [inline README.md]: Needs a `token`. Part of the SCM identifier needs to be determined by the SCM service. This requires a query against the SCM service, which is why we need a `token`

File: README.md [hunk 2]
```diff
@@ -20,23 +20,136 @@ Required parameters:
...
+    username: 'batman',
+    url: 'https://bitbucket.org/batman/test',
+    branch: 'mynewbranch',
+    sha: '40171b678527',
+    prNum: 3,
+    prRef: 'refs/pull-requests/3/from'
+}
+```
+
+#### Expected Promise response
```
Comment (d2lam) [inline README.md]: Won't be promise.

File: README.md [hunk 3]
```diff
@@ -20,23 +20,136 @@ Required parameters:
...
+| config.scmUri | String | Scm uri |
+| config.token  | String | The github token to check permissions on |
+
+#### Expected Outcome
+Decorated url in the form of:
+```js
+{
+    url: 'https://github.com/screwdriver-cd/scm-base',
+    title: 'screwdriver-cd/scm-base',
+    subtitle: 'master'
```
Comment (FenrirUnbound) [inline README.md]: How do you define another branch?

File: README.md [hunk 4]
```diff
@@ -20,23 +20,136 @@ Required parameters:
 
 | Parameter        | Type  |  Description |
 | :-------------   | :---- | :-------------|
-| scmUrl        | String | Scm Url to format |
+| scmUrl        | String | Scm url to format |
```
Comment (FenrirUnbound) [inline README.md]: `scmUri`?

Comment (FenrirUnbound) [inline README.md]: Or is this supposed to be the `checkoutUrl`?

File: README.md (hunk 1 above)
Comment (tkyi) [inline README.md]: Addressed

File: README.md (hunk 3 above)
Comment (tkyi) [inline README.md]: Addressed

File: README.md (hunk 2 above)
Comment (tkyi) [inline README.md]: Removed

File: README.md (hunk 4 above)
Comment (tkyi) [inline README.md]: Removed `formatScmUrl` function since I don't think we'll need it anymore as we are switching to `scmUri`

File: README.md [hunk 5]
```diff
@@ -15,28 +15,129 @@ This is a promise based interface for interacting with a source control manageme
...
-### formatScmUrl
+### parseUrl
 Required parameters:
 
 | Parameter        | Type  |  Description |
 | :-------------   | :---- | :-------------|
-| scmUrl        | String | Scm Url to format |
+| config             | Object | Configuration Object |
+| config.checkoutUrl | String | Checkout url for a repo to parse |
+| config.token  | String | The github token to check permissions on |
```
Comment (d2lam) [inline README.md]: Haven't implemented this for bitbucket yet, but it's not always going to be `github token`.

File: README.md [hunk 6]
```diff
@@ -15,28 +15,129 @@ This is a promise based interface for interacting with a source control manageme
...
+| payload        | Object | Request payload |
+
+#### Expected Outcome
+An object with the following fields:
+```js
+{
+    type: 'pr',         // or 'repo'
+    action: 'opened',   // or 'closed' for type 'pr'; 'push' for type 'repo'
+    username: 'batman',
+    url: 'https://bitbucket.org/batman/test.git',
```
Comment (d2lam) [inline README.md]: Sorry for wrong link earlier, this should be `https://batman@bitbucket.org/batman/test.git`

File: README.md [hunk 7]
```diff
@@ -15,28 +15,129 @@ This is a promise based interface for interacting with a source control manageme
...
+```
+
+### decorateUrl
+Required parameters:
+
+| Parameter        | Type  |  Description |
+| :-------------   | :---- | :-------------|
+| config        | Object | Configuration Object |
+| config.scmUri | String | Scm uri (ex: `github.com:1234:branchName`) |
+| config.token  | String | The github token to check permissions on |
```
Comment (d2lam) [inline README.md]: Same here, it won't be github specific.

File: README.md [hunk 8]
```diff
@@ -15,28 +15,129 @@ This is a promise based interface for interacting with a source control manageme
...
+| config.scmUri        | String | Scm uri (ex: `github.com:1234:branchName`) |
+| config.token | String | The github token to check permissions on |
+
+#### Expected Outcome
+Decorated commit in the form of:
+```js
+{
+    url: 'https://github.com/screwdriver-cd/scm-base/commit/5c3b2cc64ee4bdab73e44c394ad1f92208441411',
+    message: 'Use screwdriver to publish',
+    username: 'd2lam'
```
Comment (d2lam) [inline README.md]: Wouldn't this be a whole author instead of just username? According to data-schema 
```
  author:
 +    url: 'https://github.com/d2lam',
 +    name: 'Dao Lam',
 +    username: 'd2lam',
 +    avatar: 'https://avatars3.githubusercontent.com/u/3401924?v=3&s=400'
```

File: README.md [hunk 9]
```diff
@@ -15,28 +15,129 @@ This is a promise based interface for interacting with a source control manageme
...
+| :-------------   | :---- | :-------------|
+| headers        | Object | Request header |
+| payload        | Object | Request payload |
+
+#### Expected Outcome
+An object with the following fields:
+```js
+{
+    type: 'pr',         // or 'repo'
+    action: 'opened',   // or 'closed' for type 'pr'; 'push' for type 'repo'
```
Comment (FenrirUnbound) [inline README.md]: `synchronize` is another event that we support with Github

File: README.md [hunk 10]
```diff
@@ -103,25 +204,31 @@ The contents of the file at `path` in the repository
...
 ## Extending
 To make use of the validation functions, the functions to override are:
-1. `formatScmUrl` 
-2. `_getPermissions`
-3. `_getCommitSha`
-4. `_updateCommitStatus`
-5. `_getFile`
-6. `stats` 
+
+1. `formatScmUrl`
```
Comment (d2lam) [inline README.md]: Don't  need this now

File: index.js [hunk 11]
```diff
@@ -41,19 +41,101 @@ class ScmBase {
...
+     * @return {Object}                 Parsed hook object
+     */
+    parseHook(headers, payload) {
+        const result = this._parseHook(headers, payload);
+
+        return validate(result, dataSchema.core.scm.hook);
+    }
+
+    _parseHook() {
+        throw new Error('parseHook not implemented');
```
Comment (d2lam) [inline index.js]: Maybe just say `Not implemented` to be consistent

File: README.md (hunk 9 above)
Comment (d2lam) [inline README.md]: What is the equivalent of synchronize in bitbucket?

File: README.md (hunk 10 above)
Comment (tkyi) [inline README.md]: Addressed

File: README.md (hunk 8 above)
Comment (tkyi) [inline README.md]: Good catch, updating README

File: README.md (hunk 6 above)
Comment (tkyi) [inline README.md]: Good catch, updated

File: README.md [hunk 12]
```diff
@@ -15,28 +15,134 @@ This is a promise based interface for interacting with a source control manageme
...
+| payload        | Object | The webhook payload received from the SCM service |
+
+#### Expected Outcome
+A key-map of data related to the received payload in the form of:
+```js
+{
+    type: 'pr',         // can be 'pr' or 'repo'
+    action: 'opened',   // can be 'opened', 'closed', or 'synchronized' for type 'pr'; 'push' for type 'repo'
+    username: 'batman',
+    url: 'https://batman@bitbucket.org/batman/test.git',
```
Comment (FenrirUnbound) [inline README.md]: `checkoutUrl`

Comment (tkyi) [inline README.md]: addressed

Comment (tkyi) [inline README.md]: actually addressed

File: index.js [hunk 13]
```diff
@@ -41,19 +42,101 @@ class ScmBase {
...
+     * @param  {String}    config.scmUri    SCM uri
+     * @param  {String}    config.token     The token used to authenticate to the SCM
+     * @return {Promise}
      */
-    formatScmUrl() {
-        throw new Error('formatScmUrl not implemented');
+    decorateAuthor(config) {
+        return validate(config, dataSchema.plugins.scm.decorateAuthor)
+            .then(validAuthor => this._decorateAuthor(validAuthor))
+            .then(decoratedAuthor => validate(decoratedAuthor, dataSchema.core.scm.author));
```
Comment (FenrirUnbound) [inline index.js]: I can't find `dataSchema.core.scm.author` in https://github.com/screwdriver-cd/data-schema/blob/master/core/scm.js

Comment (tkyi) [inline index.js]: should be `dataSchema.core.scm.user`
fixed

File: index.js [hunk 14]
```diff
@@ -41,19 +42,101 @@ class ScmBase {
...
+    _decorateCommit() {
+        return Promise.reject('Not implemented');
+    }
+
+    /**
+     * Decorate the author for the specific source control
+     * @method decorateAuthor
+     * @param  {Object}    config
+     * @param  {String}    config.username  Author to decorate
+     * @param  {String}    config.scmUri    SCM uri
```
Comment (stjohnjohnson) [inline index.js]: Why is this needed?

Comment (tkyi) [inline index.js]: Removing `scmuri` as param for `decorateAuthor`

We will eventually pass in some sort of `userUri` in the form of `github.com:stjohnjohnson`, but will save that refactoring for a different pull request

File: README.md [hunk 15]
```diff
@@ -15,28 +15,133 @@ This is a promise based interface for interacting with a source control manageme
...
-### formatScmUrl
+### parseUrl
 Required parameters:
 
 | Parameter        | Type  |  Description |
 | :-------------   | :---- | :-------------|
-| scmUrl        | String | Scm Url to format |
+| config             | Object | Configuration Object |
+| config.checkoutUrl | String | Checkout url for a repo to parse |
+| config.token  | String | The scm token to check permissions on |
```
Comment (d2lam) [inline README.md]: Just say `Access token for scm` here since it's not really used "to check permissions on" anything. 
Same for other functions.

**Comment by d2lam** [comment]
The rest LGTM

File: README.md (hunk 15 above)
Comment (tkyi) [inline README.md]: changed 👍
